toolkit/components/filewatcher/NativeFileWatcherLnx.cpp

```diff
@@ -0,0 +1,1073 @@
+﻿#include "NativeFileWatcherLnx.h"
+
+#include "mozilla/Services.h"
+#include "mozilla/UniquePtr.h"
+#include "nsClassHashtable.h"
+#include "nsDataHashtable.h"
+#include "nsIFile.h"
+#include "nsIObserverService.h"
+#include "nsProxyRelease.h"
+#include "nsTArray.h"
+#include "mozilla/Logging.h"
+#include "mozilla/Scoped.h"
+
+#include <sys/inotify.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <csignal>
+#include <cstdlib>
+#include <queue>
+#include <cstring>
+#include <errno.h>
+
+#include <iostream>
+
+#include <NativeFileWatcherCommons.h>
+
+namespace mozilla {
+
+namespace moz_filewatcher {
+
+// Global signal counter for events reported by inotify. Used in event handler.
+static volatile sig_atomic_t mEventWaiting;
```

Comment:
me to review.

Reply:
While I completely understand your qualms regarding signals - it's going to be too big of a lift to switch over to a well-tested blocking read with the time we have remaining in the term (two weeks). It sounds like something a followup revision can tackle though. 

In regard to your latter questions - it's static because it's being used in a static context (the signal handler). That itself is static because a function pointer to the signal handler is required when setting up the sigaction struct, which sets up signal handling for this process. It was between a namespace global counter and signal handler versus a static member function and member variable. Either can be used, and we can switch that out if you think it's best.

Reply:
> While I completely understand your qualms regarding signals - it's going to be too big of a lift to switch over to a well-tested blocking read with the time we have remaining in the term (two weeks). It sounds like something a followup revision can tackle though.

I'm asking around, but I suspect that this is something that we would not let land, as it has chances of compromising the stability of Firefox. A few issues with signals:

- signals don't play well with memory constraints;
- signals don't play well with threads (in particular, `sig_atomic_t` is not thread-safe);
- signals don't play well with other signal handlers.

> In regard to your latter questions - it's static because it's being used in a static context (the signal handler).

This is not what `static` does in this context (thank C++ for using the same keyword for completely different uses). Here, `static` hides the variable from the linker. Which is generally good, but don't use it at random :)

> That itself is static because a function pointer to the signal handler is required when setting up the sigaction struct, which sets up signal handling for this process. It was between a namespace global counter and signal handler versus a static member function and member variable. Either can be used, and we can switch that out if you think it's best.

Either works, but I actually prefer a static member variable, as it groups data and logic in the same place.

Reply:
That all makes sense. I'll check in with our available time over the next few weeks (which is mostly blocked with presentation work and finals) and get back on whether this is feasible or not within the next few days.